File: Leetcode/2-Add-Two-Numbers/AddTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    cur1 := l1
    cur2 := l2
    dummy := &ListNode{Val : 0, Next : nil}
    cur := dummy
    carry := 0
    for cur1 != nil || cur2 != nil {
        if cur1 != nil {
            carry += cur1.Val
            cur1 = cur1.Next
        }

        if cur2 != nil {
            carry += cur2.Val
            cur2 = cur2.Next
        }

        temp := &ListNode{Val : carry%10, Next : nil}
        cur.Next = temp
        cur = cur.Next
        carry = carry / 10
    }
    if carry != 0 {
        temp := &ListNode{Val : carry, Next : nil}
        cur.Next = temp
        cur = cur.Next
    }
    return dummy.Next
}

File: Leetcode/2-Add-Two-Numbers/AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        ListNode* dummy = new ListNode(0);
        ListNode* cur = dummy;
        int carry = 0;
        while (cur1 != NULL || cur2 != NULL) {
            if (cur1 != NULL) {
                carry += cur1->val;
                cur1 = cur1->next;
            }
            if (cur2 != NULL) {
                carry += cur2->val;
                cur2 = cur2->next;
            }
            ListNode* temp = new ListNode(carry%10);
            cur->next = temp;
            cur = cur->next;
            carry = carry/10;
        }
        if (carry != 0) {
            ListNode* temp = new ListNode(carry);
            cur->next = temp;
            cur = cur->next;
        }
        return dummy->next;
    }
};


